pulp_container/app/registry_api.py

```diff
@@ -349,6 +324,24 @@ def get_dr_push(self, request, path, create=False):
                 raise RepositoryNotFound(name=path)
         return distribution, repository
 
+    def create_dr(self, path, request):
+        try:
+            with transaction.atomic():
+                repository = serializers.ContainerPushRepositorySerializer.get_or_create(
+                    {"name": path}
+                )
+                distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                    {"base_path": path, "name": path}, {"repository": get_url(repository)}
+                )
+        except ObjectDoesNotExist:
```

Comment:
this seems to be unnecessary, we whether retrieve existing instance or create a new one

Reply:
I think this is necessary. Once the validation fails (e.g., because of an invalid base_path), it tries to fetch an object from the database: https://github.com/pulp/pulpcore/blob/3594624a3556af2ad400c29871041252ae947495/pulpcore/app/serializers/base.py#L374. If the object does not exist after the last attempt, we should return `RepositoryNotFound`.

Reply:
I see what you're saying and you are right only because the get_or_create is not implemented fully as it should be. Get or create implies that on success an object will always be returned.
Here https://github.com/pulp/pulpcore/blob/3594624a3556af2ad400c29871041252ae947495/pulpcore/app/serializers/base.py#L372 we should not catch these together but separately, because if validation fails because of field being missing or invalid we need to raise it. And if it failed because 'this field must be unique; then we retrieve existing object.
Here I took a stab https://github.com/pulp/pulpcore/pull/4643